graph_def_editor/function_graph.py

```diff
@@ -0,0 +1,351 @@
+
+  def get_frame_names(self):
+    # type: () -> Tuple[str]
+    """
+    Returns:
+      Tuple of all the unique names of frames that occur in this graph.
+    """
+    if self._node_to_frame_names is None:
+      self._generate_node_to_frame_name()
+    return self._frame_name_to_nodes.keys()
+
+  def _get_op_def_denormalized_outputs(self, op):
+    # type: (Node) -> (List[op_def_pb2.OpDef.ArgDef], bool)
+    # pylint: disable=protected-access
+    op_def = self._func_graph._get_op_def(op.op_type)
+    output_args = []
+
+    input_op_output_has_number_attr = False
+    for output_arg in op_def.output_arg:
+      if output_arg.number_attr:
+        l = op.get_attr(output_arg.number_attr)
+        input_op_output_has_number_attr = True
+        for _ in range(l):
+          output_args.append(op_def_pb2.OpDef.ArgDef(name=output_arg.name,
+                                                     type=output_arg.type))
+      else:
+        output_args.append(output_arg)
+
+    return (output_args, input_op_output_has_number_attr)
+
+################################################################################
+# Stuff below this line is private to this file.
+
+
+def _get_func_graph_for_name(graph, func_name):
+  """Returns the FuncGraph and FuncDef associated to the given func_name."""
+  outer_graph = graph
+  while graph is not None:
+    # pylint: disable=protected-access
+    func = graph._get_function(str(func_name))
+    if func is not None:
+      if hasattr(func, "graph"):
+        return (func.graph, func.definition)
+      # `outer_graph` may not be the same as `ops.get_default_graph()` e.g.
+      # in the case of nested if ops or when the gradient is being computed
+      # from inside a Defun. We build the `func_graph` with `outer_graph` as its
+      # outer graph.
+      with outer_graph.as_default():
+        # This is a _DefinedFunction.
+        func_graph = (
+            function_def_to_graph.function_def_to_graph(func.definition))
+      if func_graph is not None:
+        return (func_graph, func.definition)
+    if hasattr(graph, "outer_graph"):
+      graph = graph.outer_graph
+    else:
+      raise ValueError(
+          "Function {} does not exist in the graph.".format(func_name))
+
+
+def _decode_graph(name, func_graph):
+  # type: (str, tf.Graph) -> Dict[str, List[Tuple[tf.DType, tf.TensorShape, str]]]
+  """
+  Use public TensorFlow APIs to decode the important information that is not
+  explicitly stored in the GraphDef proto, but which must be inferred from the
+  GraphDef in conjunction with additional data structures that TensorFlow
+  generally keeps to itself.
+
+  Args:
+    name: function name.
+    func_graph: tf.GraphDef protobuf that represents a function graph.
+
+  Returns:
+    A map from node name to a list of (type, shape, output_arg_name) tuples
+    that describes in turn each of the outputs of said node.
+  """
+  # The information in a NodeDef is not sufficient to determine output type
```

Comment:
This comment doesn't appear to match the code below. Would you mind replacing it with a brief explanation of what's going on in the `for` loop below? I must admit I don't fully understand all the cases in that code.

On a side note, I'm very glad I didn't have to write that `for` loop ;-)